File: macos/Sources/ElevateCore/Support/ActiveSummary.swift
import Foundation

/// Order for the "Active now" section: what expires first on top, then what is waiting.
public enum ActiveSummary {
    public static func order(_ assignments: [ActiveAssignment]) -> [ActiveAssignment] {
        let active = assignments.filter { $0.status == .active }
            .sorted { ($0.endDateTime ?? .distantFuture, $0.assignmentId ?? "") < ($1.endDateTime ?? .distantFuture, $1.assignmentId ?? "") }
        let pending = assignments.filter { $0.status == .pendingApproval }
            .sorted { ($0.startDateTime, $0.assignmentId ?? "") < ($1.startDateTime, $1.assignmentId ?? "") }
        let provisioning = assignments.filter { $0.status == .pendingProvisioning }
            .sorted { ($0.startDateTime, $0.assignmentId ?? "") < ($1.startDateTime, $1.assignmentId ?? "") }
        return active + pending + provisioning
    }
}

File: windows/src/Elevate.Core/Support/ActiveSummary.cs
using Elevate.Core.Models;

namespace Elevate.Core.Support;

/// <summary>Order for the "Active now" section: what expires first on top, then what is waiting.</summary>
public static class ActiveSummary
{
    public static IReadOnlyList<ActiveAssignment> Order(IEnumerable<ActiveAssignment> assignments)
    {
        ArgumentNullException.ThrowIfNull(assignments);
        var all = assignments.ToList();

        var active = all.Where(a => a.Status.Kind == AssignmentStatusKind.Active)
            .OrderBy(a => a.EndDateTime ?? DateTimeOffset.MaxValue)
            .ThenBy(a => a.AssignmentId ?? string.Empty, StringComparer.Ordinal);
        var scheduled = ByStart(all, AssignmentStatusKind.Scheduled);
        var pending = ByStart(all, AssignmentStatusKind.PendingApproval);
        var provisioning = ByStart(all, AssignmentStatusKind.PendingProvisioning);

        return [.. active, .. scheduled, .. pending, .. provisioning];
    }

    private static IEnumerable<ActiveAssignment> ByStart(IEnumerable<ActiveAssignment> all, AssignmentStatusKind kind) =>
        all.Where(a => a.Status.Kind == kind)
            .OrderBy(a => a.StartDateTime)
            .ThenBy(a => a.AssignmentId ?? string.Empty, StringComparer.Ordinal);
}
